src/Workspaces/Remote/Core/ExportProviderBuilder.cs

```diff
@@ -0,0 +1,220 @@
+        try
+        {
+            if (File.Exists(compositionCacheFile))
+            {
+                args.LogTrace($"Loading cached MEF catalog: {compositionCacheFile}");
+
+                CachedComposition cachedComposition = new();
+                using FileStream cacheStream = new(compositionCacheFile, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 4096, useAsync: true);
+                var exportProviderFactory = await cachedComposition.LoadExportProviderFactoryAsync(cacheStream, args.Resolver, cancellationToken).ConfigureAwait(false);
+
+                return exportProviderFactory;
+            }
+        }
+        catch (Exception ex)
+        {
+            // Log the error, and move on to recover by recreating the MEF composition.
+            args.LogError($"Loading cached MEF composition failed: {ex}");
+        }
+
+        args.LogTrace($"Composing MEF catalog using:{Environment.NewLine}{string.Join($"    {Environment.NewLine}", args.AssemblyPaths)}.");
+
+        var discovery = PartDiscovery.Combine(
+            args.Resolver,
+            new AttributedPartDiscovery(args.Resolver, isNonPublicSupported: true), // "NuGet MEF" attributes (Microsoft.Composition)
+            new AttributedPartDiscoveryV1(args.Resolver));
+
+        var parts = await discovery.CreatePartsAsync(args.AssemblyPaths, progress: null, cancellationToken).ConfigureAwait(false);
+        var catalog = ComposableCatalog.Create(args.Resolver)
+            .AddParts(parts)
+            .WithCompositionService(); // Makes an ICompositionService export available to MEF parts to import
+
+        // Assemble the parts into a valid graph.
+        var config = CompositionConfiguration.Create(catalog);
+
+        // Verify we only have expected errors.
+        ThrowOnUnexpectedErrors(config, catalog, args.ExpectedErrorParts, args.LogError);
+
+        // Try to cache the composition.
+        s_cacheWriteTask = WriteCompositionCacheAsync(compositionCacheFile, config, args.PerformCleanup, args.LogError, cancellationToken).ReportNonFatalErrorAsync();
+
+        // Prepare an ExportProvider factory based on this graph.
+        return config.CreateExportProviderFactory();
+    }
+
+    /// <summary>
+    /// Returns the path to the MEF composition cache file. Inputs used to determine the file name include:
+    /// 1) The given assembly paths
+    /// 2) The last write times of the given assembly paths
+    /// 3) The .NET runtime major version
+    /// </summary>
+    private static string GetCompositionCacheFilePath(string cacheDirectory, string catalogPrefix, ImmutableArray<string> assemblyPaths)
+    {
+        return Path.Combine(cacheDirectory, $"{catalogPrefix}.{ComputeAssemblyHash(assemblyPaths)}{CatalogSuffix}");
+
+        static string ComputeAssemblyHash(ImmutableArray<string> assemblyPaths)
+        {
+            // Ensure AssemblyPaths are always in the same order.
+            assemblyPaths = assemblyPaths.Sort();
```

Comment:
Should this be specifying some comparer? If it's accidentally going to inherit a culture-specific comparer then it might not be as stable as you might want.